src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -11,30 +12,51 @@ internal partial class Interop
     internal partial class WinHttp
     {
         // Issue 2501: Fold SafeWinHttpHandleWithCallback and SafeWinHttpHandle into a single type.
-        //
-        // SafeWinHttpHandleWithCallback is a better pattern; SafeWinHttpHandle is probably a better
-        // name. The implementation for the former should probably be folded into the latter.
-
+        // SafeWinHttpHandleWithCallback is incorrectly overriding the Dispose(bool disposing) method
+        // and will be fixed as part of merging the classes.
         internal class SafeWinHttpHandle : SafeHandleZeroOrMinusOneIsInvalid
         {
+            private SafeWinHttpHandle _parentHandle = null;
+            
             public SafeWinHttpHandle() : base(true)
             {
             }
 
-            public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
+            public static void DisposeAndClearHandle(ref SafeWinHttpHandle safeHandle)
             {
-                if (winHttpHandler != null)
+                if (safeHandle != null)
                 {
-                    winHttpHandler.Dispose();
-                    winHttpHandler = null;
+                    safeHandle.Dispose();
+                    safeHandle = null;
                 }
             }
 
+            public void SetParentHandle(SafeWinHttpHandle parentHandle)
+            {
+                Debug.Assert(_parentHandle == null);
+                Debug.Assert(parentHandle != null);
+                Debug.Assert(!parentHandle.IsInvalid);
+
+                bool success = false;
+                parentHandle.DangerousAddRef(ref success);
+                if (!success)
```

Comment:
@jkotas, the docs are just wrong (or incomplete) here, correct?  DangerousAddRef will either return successfully with the ref argument set to true or it will throw an exception leaving the original value of false in the argument, right?